src/cli/commands/add.js

```diff
@@ -55,23 +56,22 @@ export class Add extends Install {
    */
   getPatternVersion(pattern: string, pkg: Manifest): string {
     const {exact, tilde} = this.flags;
-    const parts = PackageRequest.normalizePattern(pattern);
+    const {hasVersion, range} = PackageRequest.normalizePattern(pattern);
     let version;
+
     if (getExoticResolver(pattern)) {
       // wasn't a name/range tuple so this is just a raw exotic pattern
       version = pattern;
-    } else if (parts.hasVersion && parts.range) {
+    } else if (hasVersion && range && (semver.satisfies(pkg.version, range) || getExoticResolver(range))) {
       // if the user specified a range then use it verbatim
-      version = parts.range === 'latest' ? `^${pkg.version}` : parts.range;
-    } else if (tilde) {
-      // --save-tilde
-      version = `~${pkg.version}`;
-    } else if (exact) {
-      // --save-exact
-      version = pkg.version;
+      version = range;
     } else {
-      // default to save prefix
-      version = `${String(this.config.getOption('save-prefix') || '')}${pkg.version}`;
+      const prefix = tilde
+        ? '~' // --save-tilde
+        : exact
+          ? '' // --save-exact
+          : String(this.config.getOption('save-prefix')) || '^';
+      version = `${prefix}${pkg.version}`;
```

Comment:
The previous version was incorrect since it was checking the prefix only if the `hasRange` condition was not met. That said I can convert this into if statements instead of ternaries. I also find nested ternaries harder to read.